### ISSUE TITLE
fetching from list_url no longer works when installing

### ISSUE THREAD

**Issue by glennpj** [LLNL/spack#2309]
Recent updates have broken the ability to fetch from list_url when installing. This seems to point to PR #2263 and #1822. Since the r- packages depend on list_url to fetch versions other than the latest in CRAN this has likely broken some of the spack r- packages. Here is an example with pango, which has a URL scheme that used to work with list_url.

```
class Pango(Package):                                                                                               
 29     """Pango is a library for laying out and rendering of text, with                                                
 30        an emphasis on internationalization. It can be used anywhere                                                 
 31        that text layout is needed, though most of the work on Pango so                                              
 32        far has been done in the context of the GTK+ widget toolkit."""                                              
 33     homepage = "http://www.pango.org"                                                                               
 34     url      = "http://ftp.gnome.org/pub/gnome/sources/pango/1.36/pango-1.36.8.tar.xz"                              
 35     list_url = "http://ftp.gnome.org/pub/gnome/sources/pango/"                                                      
 36     list_depth = 2                                                                                                  
 37                                                                                                                     
 38     version('1.36.8', '217a9a753006275215fa9fa127760ece')                                                           
 39     version('1.40.1', '6fc88c6529890d6c8e03074d57a3eceb')
```

```
spack versions pango
==> Safe versions (already checksummed):
  1.40.1  1.36.8
==> Remote versions (not yet checksummed):
  1.40.3  1.38.1  1.37.3  1.36.7  1.36.3  1.35.3  1.34.0  1.32.6  1.32.2  1.31.1  1.29.5
  1.40.2  1.38.0  1.37.2  1.36.6  1.36.2  1.35.2  1.33.9  1.32.5  1.32.1  1.31.0  1.29.4
  1.40.0  1.37.5  1.37.1  1.36.5  1.36.1  1.35.0  1.33.8  1.32.4  1.32.0  1.30.1  1.29.3
  1.39.0  1.37.4  1.37.0  1.36.4  1.36.0  1.34.1  1.33.7  1.32.3  1.31.2  1.30.0  1.29.1
```
```
spack stage pango
==> Fetching http://ftp.gnome.org/pub/gnome/sources/pango/1.36/pango-1.40.1.tar.xz

curl: (22) The requested URL returned error: 404 Not Found
==> Fetching from http://ftp.gnome.org/pub/gnome/sources/pango/1.36/pango-1.40.1.tar.xz failed.
==> Error: All fetchers failed for pango-1.40.1-aejizzb5ofh63jl5qvygyy6ovvcostw4
```
That used to work by going to list_url. Backing out PR #2263 gets fetching to work again to the stage but produces an error and does not put the tarball in the cache. That leads to the install failing.
```
spack stage pango
==> Fetching http://ftp.gnome.org/pub/gnome/sources/pango/1.36/pango-1.40.1.tar.xz

curl: (22) The requested URL returned error: 404 Not Found
==> Fetching from http://ftp.gnome.org/pub/gnome/sources/pango/1.36/pango-1.40.1.tar.xz failed.
==> Fetching http://ftp.gnome.org/pub/gnome/sources/pango/1.40/pango-1.40.1.tar.xz
######################################################################## 100.0%
==> Error: Attempt to check URLFetchStrategy with no digest.
```
However, a second attempt at the install will then put the previously downloaded tarball into the cache and then the installation will proceed.

@tgamblin Can you take a look at this?

**Comment by scheibelp** [comment]
Apologies, this was disabled in #2263 because when I refactored it in #1822 I inadvertently slowed down the tests.

#2270 adds it back without slowing down tests.

**Comment by omsai** [comment]
Maybe I'm misunderstanding, but with even with 85a0568 installing `pango@1.40.1` fails because the checksummed version does not seem to deal with the versioned directory in the URL:

```
curl: (22) The requested URL returned error: 404 Not Found
==> Fetching from http://ftp.gnome.org/pub/gnome/sources/pango/1.36/pango-1.40.1.tar.xz failed.
==> Error: FetchError: All fetchers failed for pango-1.40.1-jpa5bi67ssoxq2jflrjroa5srjdhculu
```

To fix that, maybe `url_for_version()` is necessary?

```python
    def url_for_version(self, version):
        return "{}{}/pango-{}.tar.xz".format(
            self.list_url, version.up_to(2), version)
```

**Comment by scheibelp** [comment]
Hi @omsai:

> Maybe I'm misunderstanding, but with even with 85a0568 installing pango@1.40.1 fails

It looks like you linked develop. As of now #2270 has not yet been merged into develop so it will still have this issue.

> To fix that, maybe url_for_version() is necessary?

This will work specifically for the pango package (I think). However, the general goal of list_url is to avoid avoid per-package logic for tracking down URLs for specific versions. 

If that doesn't clarify things or seems inconsistent let me know.

Thanks!

**Comment by adamjstewart** [comment]
Just to clarify, there are two things going on here. `list_url` may be needed to find new versions with `spack versions` and `spack checksum`, but it isn't needed to fetch packages. If you take out `list_url`, spack should still know enough to extrapolate the correct URL for the version number. This used to work.

**Comment by glennpj** [comment]
For pango, url_for_version would work. But url_for_version is not universal and will not work for r- packages. As such, list_url, **is** needed for fetching packages.

**Comment by adamjstewart** [comment]
Agreed, `list_url` is necessary for fetching R packages, but I don't think it's necessary for fetching pango. Regardless, `url_for_version` should not be necessary for pango. It wasn't necessary in the past, but something broke. `url_for_version` would fix pango, but it wouldn't fix all of the other packages that are broken. We need to identify the commit that broke it and back out any changes that are causing the problem.

**Comment by scheibelp** [comment]
Looking at the default fetcher for pango, the substituted URL is

**Comment by scheibelp** [comment]
> I don't think [list_url is] necessary for fetching pango.

Looking at the default fetcher for pango, the URL is 

```http://ftp.gnome.org/pub/gnome/sources/pango/1.36/pango-1.40.1.tar.xz ```

So list_url ends up being required (since the spidering detects the correct link while substitution of Package.url doesn't)

> list_url may be needed to find new versions with spack versions and spack checksum, but it isn't needed to fetch packages

In some cases (e.g. pango) the list_url is needed for fetching versions even if they are mentioned in the package file.

It when I refactored processing of list_url in #1822 it had two effects:

* slow down tests
* break list_url when a version was mentioned in a package.py file

#2270 fixes both of those issues

**Comment by adamjstewart** [comment]
Ok, now I'm really confused. If I checkout the commit before #1822, 4e6d5350588ab2435afa4393c75190ec0e9c031c, and edit `mpich` by commenting out the `list_url` and `list_depth`, `spack fetch` works just fine. `spack info` shows the correct URLs for each version, automatically substituting the the version multiple times in the URL. But if I try the same thing with `pango`, it does indeed require `list_url` for fetching. In fact, even with the `list_url`, it first tries the wrong URL and `spack info` shows the wrong URL. Isn't there any way we can get Spack to automatically replace the version number in the `pango` URL?

**Comment by scheibelp** [comment]
> In fact, even with the list_url, it first tries the wrong URL and spack info shows the wrong URL.

It will always try the list_url as a backup. The first thing it always attempts is to substitute into Package.url. 

> Isn't there any way we can get Spack to automatically replace the version number in the pango URL?

The url substitution logic could likely be made smarter such that pango would succeed. I dont know why it succeeds for mpich but not for pango.

> If I checkout the commit before #1822, 4e6d535, and edit mpich by commenting out the list_url and list_depth, spack fetch works just fine

Fetching mpich appears to work for now (which makes sense to me given that it doesn't need list_url)

Does clear anything up?

**Comment by glennpj** [comment]
I _think_ the reason that mpich works is because the full version is used in the directory name, whereas for pango the directory name is not the full version.

url = "http://ftp.gnome.org/pub/gnome/sources/pango/1.36/pango-1.36.8.tar.xz"
url = "http://www.mpich.org/static/downloads/3.0.4/mpich-3.0.4.tar.gz"

**Comment by adamjstewart** [comment]
Fixed by #2270 
```
$ spack fetch pango@1.36.8
==> Fetching http://ftp.gnome.org/pub/GNOME/sources/pango/1.40/pango-1.36.8.tar.xz

curl: (22) The requested URL returned error: 404 Not Found
==> Fetching from http://ftp.gnome.org/pub/GNOME/sources/pango/1.40/pango-1.36.8.tar.xz failed.
==> Fetching http://ftp.gnome.org/pub/gnome/sources/pango/1.36/pango-1.36.8.tar.xz
######################################################################## 100.0%
```
